lucene/core/src/java/org/apache/lucene/search/FilteringFieldComparator.java

```diff
@@ -0,0 +1,350 @@
+                    encodeTop(maxValueAsBytes);
+                }
+            };
+            DocIdSetBuilder result = new DocIdSetBuilder(maxDoc);
+            PointValues.IntersectVisitor visitor = new PointValues.IntersectVisitor() {
+                DocIdSetBuilder.BulkAdder adder;
+                @Override
+                public void grow(int count) {
+                    adder = result.grow(count);
+                }
+
+                @Override
+                public void visit(int docID) {
+                    if (docID <= maxDocVisited) {
+                        return; // Already visited or skipped
+                    }
+                    adder.add(docID);
+                }
+
+                @Override
+                public void visit(int docID, byte[] packedValue) {
+                    if (docID <= maxDocVisited) {
+                        return;  // Already visited or skipped
+                    }
+                    if (maxValueExist) {
+                        // doc's value is too high
+                        if (Arrays.compareUnsigned(packedValue, 0, bytesCount, maxValueAsBytes, 0, bytesCount) > 0) return;
+                    }
+                    if (minValueExist) {
+                        // doc's value is too low
+                        if (Arrays.compareUnsigned(packedValue, 0, bytesCount, minValueAsBytes, 0, bytesCount) < 0) return;
+                    }
+                    adder.add(docID); // doc is competitive
+                }
+
+                @Override
+                public PointValues.Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
+                    if ((maxValueExist && Arrays.compareUnsigned(minPackedValue, 0, bytesCount, maxValueAsBytes, 0, bytesCount) > 0) ||
+                            (minValueExist && Arrays.compareUnsigned(maxPackedValue, 0, bytesCount, minValueAsBytes, 0, bytesCount) < 0)) {
+                        return PointValues.Relation.CELL_OUTSIDE_QUERY;
+                    }
+                    if ((maxValueExist && Arrays.compareUnsigned(maxPackedValue, 0, bytesCount, maxValueAsBytes, 0, bytesCount) > 0) ||
+                            (minValueExist && Arrays.compareUnsigned(minPackedValue, 0, bytesCount, minValueAsBytes, 0, bytesCount) < 0)) {
+                        return PointValues.Relation.CELL_CROSSES_QUERY;
+                    }
+                    return PointValues.Relation.CELL_INSIDE_QUERY;
+                }
+            };
+            final long threshold = iterator.cost() >>> 3;
+            long estimatedNumberOfMatches = pointValues.estimatePointCount(visitor); // runs in O(log(numPoints))
+            if (estimatedNumberOfMatches >= threshold) {
+                // the new range is not selective enough to be worth materializing, it doesn't reduce number of docs at least 8x
+                return;
+            }
+            pointValues.intersect(visitor);
+            this.iterator = result.build().iterator();
```

Comment:
I believe that doing it is incorrect if a missing value is configured and would be competitive.

Reply:
addressed in  24c94ff